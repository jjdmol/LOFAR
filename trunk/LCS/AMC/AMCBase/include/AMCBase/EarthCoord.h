//# EarthCoord.h: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_AMCBASE_EARTHCOORD_H
#define LOFAR_AMCBASE_EARTHCOORD_H

// \file EarthCoord.h
// Class to hold an earth coordinate as lon,lat,height

//# Forward Declarations.
#include <Common/lofar_iosfwd.h>
#include <Common/lofar_string.h>

namespace LOFAR
{
  namespace AMC
  {

    // \addtogroup AMCBase
    // @{

    // This class represents a position on earth. The position is stored using
    // longitude, latitude and height. The context where the object is used
    // defines the coordinate system and frame, so the class can be used in
    // any kind of frame (like ITRF and geocentric). The correct
    // interpretation of the coordinates should be done by the user of this
    // class.
    class EarthCoord
    {
    public:
      // Types of earth coordinates. 
      enum Types {
        INVALID = -1,   ///< Used when specified value is out of range.
        ITRF,
        WGS84,
        //# Insert new types HERE !!
        N_Types         ///< Number of reference types.
      };

      // Default constructor uses 0 for the values.
      EarthCoord()
        : itsLong(0), itsLat(0), itsHeight(0), itsType(ITRF) {}

      // Create a earth coordinate by giving the longitude and latitude in
      // radians and the height in meters. Reference type can be either ITRF
      // (default), or WGS84.
      EarthCoord (double longitude, double latitude, double height=0,
                  Types typ = ITRF);

      // Return the longitude in radians.
      double longitude() const
      { return itsLong; }

      // Return the latitude in radians.
      double latitude() const
      { return itsLat; }

      // Return the height in meters.
      double height() const
      { return itsHeight; }

      // Return the reference type.
      Types type() const
      { return itsType; }

      // Return the reference type as a string.
      const string& showType() const;

      // Return whether sky coordinate type is valid.
      bool isValid() const
      { return itsType != INVALID; }

    private:
      // Longitude in radians.
      double itsLong;

      // Latitude in radians.
      double itsLat;

      // Height in meters.
      double itsHeight;

      // Type of earth coordinate.
      Types itsType;
    };

    // Output an EarthCoord in ASCII format.
    ostream& operator<< (ostream&, const EarthCoord&);

    // Compare two EarthCoord objects for equality.
    // \note Two invalid objects can \e never be equal.
    bool operator==(const EarthCoord& lhs, const EarthCoord& rhs);

    // @}

  } // namespace AMC

} // namespace LOFAR

#endif
