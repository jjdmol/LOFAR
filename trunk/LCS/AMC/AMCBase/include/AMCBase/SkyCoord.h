//# SkyCoord.h: Class to hold a sky coordinate as 2 angles
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef AMC_AMCBASE_SKYCOORD_H
#define AMC_AMCBASE_SKYCOORD_H

//# Forward Declarations.
#include <Common/lofar_iosfwd.h>

namespace LOFAR
{
  namespace AMC
  {

    class SkyCoord
    {
    public:
      // Default constructor uses 0 for the angles.
      SkyCoord()
        : itsAngle0(0), itsAngle1(0) {}

      // Create a sky coordinate by giving the longitude and latitude in
      // radians.  The context where the object is used defines the coordinate
      // system and frame, so the class can be used for any pair of sky
      // coordinates (like RA/DEC and AZ/ELEV).
      SkyCoord (double angle0, double angle1)
        : itsAngle0(angle0), itsAngle1(angle1) {}

      // Get the values out.
      double angle0() const
      { return itsAngle0; }
      double angle1() const
      { return itsAngle1; }

    private:
      double itsAngle0;
      double itsAngle1;
    };

    // Output in ASCII.
    ostream& operator<< (ostream&, const SkyCoord&);

  } // namespace AMC

} // namespace LOFAR


#endif
