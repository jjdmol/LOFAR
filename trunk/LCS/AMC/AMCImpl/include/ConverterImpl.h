//#  ConverterImpl.h: implementation of the Converter interface.
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_AMCIMPL_CONVERTERIMPL_H
#define LOFAR_AMCIMPL_CONVERTERIMPL_H

//# Never #include <config.h> or #include <lofar_config.h> in a header file!

//# Includes
#include <AMCBase/Converter.h>

namespace LOFAR
{
  namespace AMC
  {
    // This class represents the client side of the client/server
    // implementation of the AMC. It implements the Converter interface.
    class ConverterImpl : public Converter
    {
    public:
      ConverterImpl();

      virtual ~ConverterImpl() {}

      virtual SkyCoord j2000ToAzel(const SkyCoord& radec, 
                                   const EarthCoord& pos, 
                                   const TimeCoord& time);

      virtual vector<SkyCoord> j2000ToAzel (const vector<SkyCoord>& radec,
                                            const EarthCoord& pos,
                                            const TimeCoord& time);

      virtual vector<SkyCoord> j2000ToAzel (const SkyCoord& radec,
                                            const vector<EarthCoord>& pos,
                                            const TimeCoord& time);

      virtual vector<SkyCoord> j2000ToAzel (const SkyCoord& radec,
                                            const EarthCoord& pos,
                                            const vector<TimeCoord>& time);

      virtual vector<SkyCoord> j2000ToAzel (const vector<SkyCoord>& radec,
                                            const vector<EarthCoord>& pos,
                                            const vector<TimeCoord>& time);

      virtual SkyCoord azelToJ2000 (const SkyCoord& azel,
                                    const EarthCoord& pos,
                                    const TimeCoord& time);

      virtual vector<SkyCoord> azelToJ2000 (const vector<SkyCoord>& azel,
                                            const EarthCoord& pos,
                                            const TimeCoord& time);

      virtual vector<SkyCoord> azelToJ2000 (const vector<SkyCoord>& azel,
                                            const vector<EarthCoord>& pos,
                                            const vector<TimeCoord>& time);
    private:

    };

  } // namespace AMC
  
} // namespace LOFAR

#endif
