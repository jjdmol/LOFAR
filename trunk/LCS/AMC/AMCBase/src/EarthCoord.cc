//# EarthCoord.cc: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/EarthCoord.h>
#include <Common/lofar_iostream.h>
#include <cmath>

namespace LOFAR
{
  namespace AMC
  {

    EarthCoord::EarthCoord (double longitude, double latitude, double height,
                            Types typ)
      : itsLong(longitude), itsLat(latitude), itsHeight(height)
    {
      if (INVALID < typ && typ < N_Types) itsType = typ; 
      else itsType = INVALID;
    }


    const string& EarthCoord::showType() const
    {
      //# Caution: Always keep this array of strings in sync with the enum
      //#          Types that is defined in the header file!
      static const string types[N_Types+1] = {
        "ITRF",
        "WGS84",
        "<INVALID>"
      };
      if (isValid()) return types[itsType];
      else return types[N_Types];
    }


    vector<double> EarthCoord::xyz() const
    {
      vector<double> p(3);
      double tmp = std::cos(itsLat);
      p[0] = itsHeight * std::cos(itsLong) * tmp;
      p[1] = itsHeight * std::sin(itsLong) * tmp;
      p[2] = itsHeight * std::sin(itsLat);
      return p;
    }


    ostream& operator<<(ostream& os, const EarthCoord& pos)
    {
      if (!pos.isValid()) 
        return os << pos.showType();
      else 
        return os << "["  << pos.longitude() << ", " << pos.latitude() 
                  << ", " << pos.height() << "] (" << pos.showType() << ")";
    }


    bool operator==(const EarthCoord& lhs, const EarthCoord& rhs)
    {
      return 
        lhs.isValid()   && rhs.isValid()   &&
        lhs.longitude() == rhs.longitude() &&
        lhs.latitude()  == rhs.latitude()  &&
        lhs.height()    == rhs.height()    &&
        lhs.type()      == rhs.type();
    }

  } // namespace AMC

} // namespace LOFAR
