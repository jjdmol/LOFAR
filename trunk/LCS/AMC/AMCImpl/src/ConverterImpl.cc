//#  ConverterImpl.cc: one line description
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCImpl/ConverterImpl.h>
#include <AMCImpl/Exceptions.h>
#include <AMCBase/SkyCoord.h>
#include <AMCBase/EarthCoord.h>
#include <AMCBase/TimeCoord.h>
#include <Common/LofarLogger.h>
#include <measures/Measures/MDirection.h>
#include <measures/Measures/MEpoch.h>
#include <measures/Measures/MPosition.h>
#include <measures/Measures/MeasConvert.h>
#include <casa/Exceptions/Error.h>

using namespace casa;

namespace LOFAR
{
  namespace AMC
  {

    ConverterImpl::ConverterImpl()
    {
    }


    SkyCoord 
    ConverterImpl::j2000ToAzel (const SkyCoord& skyCoord, 
                                const EarthCoord& earthCoord, 
                                const TimeCoord& timeCoord)
    {
      return j2000ToAzel (vector<SkyCoord>  (1, skyCoord),
                          vector<EarthCoord>(1, earthCoord),
                          vector<TimeCoord> (1, timeCoord)) [0];
    }


    vector<SkyCoord> 
    ConverterImpl::j2000ToAzel (const vector<SkyCoord>& skyCoord,
                                const EarthCoord& earthCoord,
                                const TimeCoord& timeCoord)
    {
      return j2000ToAzel(skyCoord, 
                         vector<EarthCoord>(1, earthCoord),
                         vector<TimeCoord> (1, timeCoord));
    }


    vector<SkyCoord> 
    ConverterImpl::j2000ToAzel (const SkyCoord& skyCoord,
                                const vector<EarthCoord>& earthCoord,
                                const TimeCoord& timeCoord)
    {
      return j2000ToAzel(vector<SkyCoord> (1, skyCoord),
                         earthCoord, 
                         vector<TimeCoord>(1, timeCoord));
    }


    vector<SkyCoord> 
    ConverterImpl::j2000ToAzel (const SkyCoord& skyCoord,
                                const EarthCoord& earthCoord,
                                const vector<TimeCoord>& timeCoord)
    {
      return j2000ToAzel(vector<SkyCoord>  (1, skyCoord),
                         vector<EarthCoord>(1, earthCoord),
                         timeCoord);
    }


    vector<SkyCoord> 
    ConverterImpl::j2000ToAzel (const vector<SkyCoord>& skyCoord,
                                const vector<EarthCoord>& earthCoord,
                                const vector<TimeCoord>& timeCoord)
    {
      // This vector will hold the result of the conversion operation.
      vector<SkyCoord> result;

      try {

        // Create a container for the Measure frame.
        MeasFrame frame;

        // Create a dummy(?) astronomical direction. We (only) need it as
        // argument for the MDirection::Convert constructor.
        MDirection dir;

        // For each given moment in time ...
        for (uint i = 0; i < timeCoord.size(); i++) {

          // Set the instant in time in the frame.
          frame.set (MEpoch(MVEpoch(timeCoord[i].getDay(), 
                                    timeCoord[i].getFraction())));

          // For each given position on earth ...
          for (uint j = 0; j < earthCoord.size(); j++) {

            // Set the position on earth in the frame.
            MVPosition pos((Quantity(earthCoord[j].height(), "m")),
                           (Quantity(earthCoord[j].longitude(), "rad")),
                           (Quantity(earthCoord[j].latitude(), "rad")));
            frame.set (MPosition(pos));

            // Set-up the conversion engine, using reference direction AZEL.
            // \note Since we're only using \a dir as a dummy argument, we
            // might as well use a temporary MDirection object.
            MDirection::Convert 
              conv (dir, MDirection::Ref (MDirection::AZEL, frame));
//            conv (MDirection(), MDirection::Ref (MDirection::AZEL, frame));
            
            // For each given direction in the sky ...
            for (uint k = 0; k < skyCoord.size(); k++) {

              // Define the astronomical direction as a J2000 direction.
              MDirection sky((MVDirection(skyCoord[k].angle0(), 
                                          skyCoord[k].angle1())), 
                             MDirection::J2000);

              // Convert this direction, using the conversion engine.
              MDirection res = conv(sky);

              // Retrieve the direction angles of the converted direction.
              Vector<Double> angles = res.getAngle().getBaseValue();

              // Convert to local sky coordinates and add to the return vector.
              result.push_back(SkyCoord(angles(0), angles(1)));

            }
          }
        }
      }

      catch (AipsError& e) {
        THROW (ConverterError, "AipsError: " << e.what());
      }

      // Check on post-condition that the return vector contains 
      // <tt>skyCoord.size() * earthCoord.size() * timeCoord.size()</tt>
      // elements.
      ASSERT (result.size() == 
              skyCoord.size() * earthCoord.size() * timeCoord.size());

      return result;
    }


    SkyCoord 
    ConverterImpl::azelToJ2000 (const SkyCoord& skyCoord,
                                const EarthCoord& earthCoord,
                                const TimeCoord& timeCoord)
    {
      return azelToJ2000(vector<SkyCoord>  (1, skyCoord),
                         vector<EarthCoord>(1, earthCoord),
                         vector<TimeCoord> (1, timeCoord)) [0];
    }


    vector<SkyCoord>
    ConverterImpl::azelToJ2000 (const vector<SkyCoord>& skyCoord,
                                const EarthCoord& earthCoord,
                                const TimeCoord& timeCoord)
    {
      return azelToJ2000(skyCoord,
                         vector<EarthCoord>(1, earthCoord),
                         vector<TimeCoord> (1, timeCoord));
    }


    vector<SkyCoord>
    ConverterImpl::azelToJ2000 (const vector<SkyCoord>& skyCoord,
                                const vector<EarthCoord>& earthCoord,
                                const vector<TimeCoord>& timeCoord)
    {
      // Check pre-conditions.
      // -# \a earthCoord and \a timeCoord must have equal sizes;
      // -# if \a earthCoord and \a timeCoord have sizes unequal to one, then
      //    their sizes must be equal to the size of \a skyCoord. Note that
      //    we only need to do one check, the other is implied by the first
      //    pre-condition.
      ASSERT (earthCoord.size() == timeCoord.size());
      if (earthCoord.size() != 1) {
        ASSERT (earthCoord.size() == skyCoord.size());
      }

      // This vector will hold the result of the conversion operation.
      vector<SkyCoord> result;

      try {

        // Create a container for the Measure frame.
        MeasFrame frame;

        // Create a dummy(?) astronomical direction. We (only) need it as
        // argument for the MDirection::Convert constructor.
        MDirection dir;

        // If there's only one earthCoord and one timeCoord, we can preset
        // these data in the Measure frame; we only need to loop over all 
        // \a skyCoord values.
        if (timeCoord.size() == 1) {

          // Set the instant in time in the frame.
          frame.set (MEpoch(MVEpoch(timeCoord[0].getDay(), 
                                    timeCoord[0].getFraction())));

          // Set the position on earth in the frame.
          MVPosition pos((Quantity(earthCoord[0].height(), "m")),
                         (Quantity(earthCoord[0].longitude(), "rad")),
                         (Quantity(earthCoord[0].latitude(), "rad")));
          frame.set (MPosition(pos));

          // Set-up the reference frame, using reference direction AZEL.
          MDirection::Ref ref(MDirection::AZEL, frame);

          // Set-up the conversion engine, using reference direction J2000..
          MDirection::Convert conv (dir, MDirection::J2000);
//          conv (MDirection(), MDirection::Ref (MDirection::J2000, frame));
            
          // For each given direction in the sky ...
          for (uint i = 0; i < skyCoord.size(); i++) {
            
            // Define the astronomical direction w.r.t. the reference frame.
            MDirection sky((MVDirection(skyCoord[i].angle0(), 
                                        skyCoord[i].angle1())), 
                           ref);
            
            // Convert this direction, using the conversion engine.
            MDirection res = conv(sky);

            // Retrieve the direction angles of the converted direction.
            Vector<Double> angles = res.getAngle().getBaseValue();

            // Convert to local sky coordinates and add to the return vector.
            result.push_back(SkyCoord(angles(0), angles(1)));
            
          }

        }
        // We need to calculate the converted sky coordinate for each triplet
        // of \a skyCoord, \a earthCoord, and \a timeCoord.
        else {

          // For each triplet ...
          for (uint i = 0; i < skyCoord.size(); i++) {

            // Set the instant in time in the frame.
            frame.set (MEpoch(MVEpoch(timeCoord[i].getDay(), 
                                      timeCoord[i].getFraction())));
            
            // Set the position on earth in the frame.
            MVPosition pos((Quantity(earthCoord[i].height(), "m")),
                           (Quantity(earthCoord[i].longitude(), "rad")),
                           (Quantity(earthCoord[i].latitude(), "rad")));
            frame.set (MPosition(pos));

            // Set-up the reference frame, using reference direction AZEL.
            MDirection::Ref ref(MDirection::AZEL, frame);
            
            // Set-up the conversion engine, using reference direction J2000..
            MDirection::Convert conv (dir, MDirection::J2000);
//          conv (MDirection(), MDirection::Ref (MDirection::J2000, frame));
           
            // Define the astronomical direction w.r.t. the reference frame.
            MDirection sky((MVDirection(skyCoord[i].angle0(), 
                                        skyCoord[i].angle1())),
                           ref);
            
            // Convert this direction, using the conversion engine.
            MDirection res = conv(sky);

            // Retrieve the direction angles of the converted direction.
            Vector<Double> angles = res.getAngle().getBaseValue();

            // Convert to local sky coordinates and add to the return vector.
            result.push_back(SkyCoord(angles(0), angles(1)));
            
          }

        }

      }
      catch (AipsError& e) {
        THROW (ConverterError, "AipsError: " << e.what());
      }

      // Check post-condition.
      ASSERT(result.size() == skyCoord.size());

      // Return the result vector.
      return result;
    }
    
    
  } // namespace AMC
  
} // namespace LOFAR
