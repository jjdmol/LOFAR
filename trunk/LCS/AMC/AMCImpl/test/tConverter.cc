//#  tConverter.cc: test program for the AMC converter classes.
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/SkyCoord.h>
#include <AMCBase/EarthCoord.h>
#include <AMCBase/TimeCoord.h>
#include <AMCBase/AMCClient/ConverterClient.h>
#include <AMCImpl/ConverterImpl.h>
#include <Common/LofarLogger.h>
#include <Common/Exception.h>
#include <Common/LofarTypes.h>
#include <Common/lofar_fstream.h>

using namespace LOFAR;
using namespace LOFAR::AMC;

// Maximum allowable error (absolute or relative).
const double epsilon = 1.2e-12;


bool compare(double x, double y, 
             double eps = std::numeric_limits<double>::epsilon())
{
  if (x == y) return true;
  if (fabs(x-y) <= eps) return true;
  if (fabs(x) > fabs(y)) return (fabs((x-y)/x) <= eps);
  else return (fabs((x-y)/y) <= eps);
}


template<typename T>
ostream& operator<<(ostream& os, const vector<T>& v)
{
  for (uint i=0; i<v.size(); i++) {
    os << endl << "  [" << i << "]: " << v[i];
  }
  return os;
}


Converter* createConverter()
{
  ifstream ifs("tConverter.in");
  if (!ifs) {
    LOG_ERROR("Failed to open file \"tConverter.in\" for reading");
    return 0;
  }

  string type;
  string host("localhost");
  uint16 port(31337);

  ifs >> type >> host >> port;

  if (type == "impl" || type == "IMPL" || type == "Impl") 
    return new ConverterImpl();
  if (type == "client" || type == "CLIENT" || type == "Client")
    return new ConverterClient(host, port);

  LOG_ERROR_STR("Invalid converter type (" << type << ") specified");
  return 0;
}


int main(int /*argc*/, const char* const argv[])
{
  INIT_LOGGER(argv[0]);

  try {

    Converter* conv = createConverter();
    if (conv == 0) return 1;
    
    vector<SkyCoord>   skies(4);
    vector<EarthCoord> poss(2);
    vector<TimeCoord>  times(3);

    // Four arbitrary directions (angles in radians).
    skies[0] = SkyCoord (1, 0.3);
    skies[1] = SkyCoord (-1, -0.3);
    skies[2] = SkyCoord (2, 0.1);
    skies[3] = SkyCoord (1.5, -0.3);

    poss[0] =  // arbitrary position
      EarthCoord (0.1, 0.8, 1000, EarthCoord::WGS84);
    poss[1] =  // Dwingeloo observatory
      EarthCoord (0.111646531, 0.921760253, 25, EarthCoord::WGS84);

    times[0] = // arbitrary time (added 0.1 usec to avoid rouding errors)
      TimeCoord (2004, 11, 19, 15, 22, 39.0000001);
    times[1] = // same arbitrary time + 1 hour
      TimeCoord (times[0].mjd(), 1./24);
    times[2] = // same arbitrary time + 1 day
      TimeCoord (times[0].mjd(), 1.);

    cout.precision(9);

    cout << "**** Original data ****" << endl;
    cout << "skies = " << skies << endl;
    cout << "poss = " << poss << endl;
    cout << "times = " << times << endl;


    // Check conversion J2000 --> AZEL and vice versa
    {
      vector<SkyCoord> result = conv->j2000ToAzel (skies, poss, times);
      ASSERT (result.size() == skies.size() * poss.size() * times.size());
      cout << "\n**** Convert from J2000 to AZEL and vice versa ****\n";
      cout << "result = " << result << endl;
      for (uint i = 0; i < skies.size() * poss.size() * times.size(); i++) {
        ASSERT (result[i].type() == SkyCoord::AZEL);
      }
      int inx=0;
      for (uint i = 0; i < times.size(); i++) {
        for (uint j = 0; j < poss.size(); j++) {
          vector<SkyCoord> sk(skies.size());
          for (uint k = 0; k < skies.size(); k++) {
            sk[k] = result[inx++];
          }
          // Convert back from AZEL to J2000; should yield original data
          vector<SkyCoord> res2 = conv->azelToJ2000 (sk, poss[j], times[i]);
          for (uint k = 0; k < skies.size(); k++) {
            ASSERT (res2[k].type() == SkyCoord::J2000);
            ASSERT (compare(res2[k].angle0(), skies[k].angle0(), epsilon));
            ASSERT (compare(res2[k].angle1(), skies[k].angle1(), epsilon));
          }
        }
      }
    }    

    // Check conversion J2000 --> ITRF 
    {
      vector<SkyCoord> result = conv->j2000ToItrf (skies, poss, times);
      ASSERT (result.size() == skies.size() * poss.size() * times.size());
      cout << "\n**** Convert from J2000 to ITRF and vice versa ****\n";
      cout << "result = " << result << endl;
      for (uint i = 0; i < skies.size() * poss.size() * times.size(); i++) {
        ASSERT (result[i].type() == SkyCoord::ITRF);
      }
      int inx=0;
      for (uint i = 0; i < times.size(); i++) {
        for (uint j = 0; j < poss.size(); j++) {
          vector<SkyCoord> sk(skies.size());
          for (uint k = 0; k < skies.size(); k++) {
            sk[k] = result[inx++];
          }
          // Convert back from ITRF to J2000; should yield original data
          vector<SkyCoord> res2 = conv->itrfToJ2000 (sk, poss[j], times[i]);
          for (uint k = 0; k < skies.size(); k++) {
            ASSERT (res2[k].type() == SkyCoord::J2000);
            ASSERT (compare(res2[k].angle0(), skies[k].angle0(), epsilon));
            ASSERT (compare(res2[k].angle1(), skies[k].angle1(), epsilon));
          }
        }
      }
    }    

  } catch (Exception& e) {
    LOG_ERROR_STR(e);
    return 1;
  }

  cout << "OK" << endl;
  return 0;
  
}
