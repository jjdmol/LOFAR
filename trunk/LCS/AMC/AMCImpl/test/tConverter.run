#!/bin/sh

# Use LimitExec if available. Let the program run for at most 20 seconds.
if [ -x $LOFARROOT/bin/LimitExec ]
then
  EXEC="$LOFARROOT/bin/LimitExec 20"
elif [ -x $lofar_sharedir/limitexec.sh ]
then
  EXEC="$lofar_sharedir/limitexec.sh 20"
fi

# The type of test to run is specified in the file tConverter.in.
# If type is "client" the third (optional) argument is the port number
INFILE="`dirname $0`/tConverter.in"
if [ -f $INFILE ]
then
   TYPE=`head -1 $INFILE | cut -d' ' -f1 | tr '[A-Z]' '[a-z]'`
   PORT=`head -1 $INFILE | cut -d' ' -f3`
fi

# Variable to keep track of the return status.
STATUS=0

# If we must run a client/server test, we must first start the server process.
if [ "$TYPE" = "client" ]
then

  # This is the AMC server executable
  SERVER="`dirname $0`/../src/amcserver"

  # The AMC server process needs its own log_prop file. We will copy 
  # tConverter.log_prop so that the same logging levels are set.
  if [ -f tConverter.log_prop ]
  then
    sed -e "s/tConverter_tmp/amcserver_tmp/" \
      tConverter.log_prop > $SERVER.log_prop
  fi

  # Exit and interrupt handler. $! is the process id of the last background 
  # process started, i.e. the converter server.
  trap 'if [ $STATUS != 0 ]; then \
          mv amcserver_tmp.log amcserver.err 2>/dev/null; \
        else \
          rm -f amcserver_tmp.log; \
        fi; \
        if kill -0 $! 2>/dev/null; then kill -9 $! 2>/dev/null; fi; \
        exit $STATUS' 0 1 2 3 15

  # Start the server in the background.
  $SERVER $PORT &

  # Sleep for 2 seconds to make sure the server has started.
  sleep 2

fi

# Start the test program.
$EXEC ./tConverter
STATUS=$?

# If something went wrong, secure the log file(s)
if [ $STATUS != 0 ]
then
  mv tConverter_tmp.log tConverter_$TYPE.err 2>/dev/null
fi

exit $STATUS
