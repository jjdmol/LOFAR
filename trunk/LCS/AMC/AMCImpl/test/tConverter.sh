#!/bin/sh
echo "impl" > tConverter.in
$lofar_sharedir/runtest.sh -stdout tConverter > tConverter_impl.log
echo "client localhost 31339" > tConverter.in
$lofar_sharedir/runtest.sh -stdout tConverter > tConverter_client.log
