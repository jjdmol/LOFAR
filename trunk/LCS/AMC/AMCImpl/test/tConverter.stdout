[1, 0.4] (J2000)
[0.2, 1, 3] (ITRF)
53328.6
2004-11-19 15:22:39.000000
2004-11-19 16:22:39.000000
2004-11-20 15:22:39.000000
[0.905774, 0.0210954] (AZEL)
[1, 0.4] (J2000)
[0.958474, -0.140808] (AZEL)
[3.00374, 0.464814] (AZEL)
[0.0160604, -0.670867] (AZEL)
[0.875086, -0.913949] (AZEL)
[0.9575, -0.0645041] (AZEL)
[3.02252, 0.345026] (AZEL)
[0.028332, -0.54897] (AZEL)
[0.783431, -0.825946] (AZEL)
[1.15285, 0.018226] (AZEL)
[-2.99878, 0.46436] (AZEL)
[0.343267, -0.638065] (AZEL)
[1.16962, -0.757774] (AZEL)
[1.16671, 0.0738909] (AZEL)
[-2.99425, 0.342781] (AZEL)
[0.33045, -0.520622] (AZEL)
[1.08321, -0.698448] (AZEL)
[0.971747, -0.130955] (AZEL)
[3.02206, 0.466352] (AZEL)
[0.0379042, -0.670544] (AZEL)
[0.897162, -0.904667] (AZEL)
[0.971562, -0.0559598] (AZEL)
[3.03995, 0.34617] (AZEL)
[0.0483872, -0.548571] (AZEL)
[0.80498, -0.818531] (AZEL)
[1, 0.3] (J2000); [-1, -0.3] (J2000); [2, 0.1] (J2000); [1.5, -0.3] (J2000); 
[1, 0.3] (J2000); [-1, -0.3] (J2000); [2, 0.1] (J2000); [1.5, -0.3] (J2000); 
[1, 0.3] (J2000); [-1, -0.3] (J2000); [2, 0.1] (J2000); [1.5, -0.3] (J2000); 
[1, 0.3] (J2000); [-1, -0.3] (J2000); [2, 0.1] (J2000); [1.5, -0.3] (J2000); 
[1, 0.3] (J2000); [-1, -0.3] (J2000); [2, 0.1] (J2000); [1.5, -0.3] (J2000); 
[1, 0.3] (J2000); [-1, -0.3] (J2000); [2, 0.1] (J2000); [1.5, -0.3] (J2000); 
OK
